secp256kfun/src/point.rs

```diff
@@ -114,6 +114,175 @@ impl Point<Normal, Public, NonZero> {
+    /// uniform distribution with a simple modification.
+    ///
+    /// [`hash_to_curve_rfc9381_tai`]: Self::hash_to_curve_rfc9381_tai
+    ///
+    /// # Example
+    /// ```
+    /// # use secp256kfun::{Point, hash::{Hash32, HashAdd}};
+    /// # use sha2::Sha256;
+    /// let hasher = Sha256::default().add(b"hello world");
+    /// let point = Point::hash_to_curve(hasher);
+    /// ```
+    ///
+    /// [`hash_to_curve_sswu`]: Self::hash_to_curve_sswu
+    /// [RFC 9381]: https://datatracker.ietf.org/doc/html/rfc9381
+    pub fn hash_to_curve<H: Hash32>(hasher: H) -> Point<Normal, Public, NonZero> {
+        use crate::hash::HashAdd;
+
+        // Try up to 255 times (probability of failure is negligible)
+        for counter in 0u8..u8::MAX {
+            let hash_bytes = hasher.clone().add(counter).finalize_fixed();
+
+            // Use 0x03 (odd y) when counter==0, 0x02 (even y) when counter>0
+            // This ensures uniform distribution over all curve points because there is
+            // a roughly 50% chance that counter will be 0 when we succeed, and this
+            // probability is independent of the x coordinate distribution
+            let mut bytes = [0u8; 33];
+            bytes[0] = 0x02 + (counter > 0) as u8;
+            bytes[1..].copy_from_slice(&hash_bytes);
+
+            if let Some(point) = Point::<Normal, Public, NonZero>::from_bytes(bytes) {
+                return point;
+            }
+        }
+
+        // This should never happen (probability ~ 2^-128)
+        unreachable!("Failed to find valid point after 128 attempts")
+    }
+
+    /// Hash to curve using [RFC 9381] try-and-increment format
+    ///
+    /// This implements a hash-to-curve method following [RFC 9381]'s try-and-increment
+    /// algorithm as used in SECP256K1_SHA256_TAI. Note that SECP256K1_SHA256_TAI is not
+    /// defined in the RFC itself, but is a ciphersuite adopted by various VRF implementations.
+    ///
+    /// This method always produces points with even y-coordinates (0x02 prefix) which means it's
+    /// not quite uniform (but this is not a security problem in any reasonable protocol)
+    ///
+    /// Like other try-and-increment methods, this has variable runtime based on input.
+    /// See [`hash_to_curve_sswu`] for discussion of DoS considerations.
+    ///
+    /// [RFC 9381]: https://datatracker.ietf.org/doc/html/rfc9381#section-5.4.1.1
+    ///
+    /// # Example
+    /// ```
+    /// # use secp256kfun::Point;
+    /// # use sha2::Sha256;
+    /// let point = Point::hash_to_curve_rfc9381_tai::<Sha256>(b"hello world", b"my-salt");
+    /// // Use empty bytes if no salt is needed
+    /// let point2 = Point::hash_to_curve_rfc9381_tai::<Sha256>(b"hello world", b"");
+    /// ```
+    ///
+    /// [`hash_to_curve_sswu`]: Self::hash_to_curve_sswu
+    pub fn hash_to_curve_rfc9381_tai<H: Hash32>(
+        msg: &[u8],
+        salt: &[u8],
+    ) -> Point<Normal, Public, NonZero> {
```

Comment:
This should return an `EvenY` point if what the documentation says is correct. It should use `from_xonly_bytes.`